### PR TITLE
Develop stream 2024-07-01

### PR DESCRIPTION
**Pull request by Naraenda**
This PR brings various updates, intended for ROCm 6.3.

It contains the following merge commits:

* 0350e0f5 perf: improve performance on rocm 6.2.
* ade3f6b2 fix: performance regression introduced with hip graph support.
* c90e74c1 docs: add documentation for hip graph support
* fa13dd77 feat: hip graph support for poisson distribution
* 4224428f feat: host implementation for mt19937
* 615fc5a7 perf: use alias method instead of binary search in cdf
* 8f7e2af7 docs: reference `GPU_TARGETS` instead of `AMDGPU_TARGET`
* 68072698 fix: fix host generators not always supporting large sizes
* 50dab4f9 refactor: remove deprecated internal headers
* 80c16d40 test: improve accuracy of poisson histogram test
* 1a44b48c docs: fix styling

This PR does not contain fixes for the recent performance regressions, we'll either add them here or create a new PR for those.

### CODE REVIEW

**Comment by Naraenda** [comment]
@samjwu The clang format CI step seems to not handle external PRs properly, I think this might also be happening in other repositories.

**Comment by Naraenda** [comment]
@stanleytsang-amd Let me know if I should squash the commits into their respective merge commits.

**Comment by stanleytsang-amd** [comment]
> @samjwu The clang format CI step seems to not handle external PRs properly, I think this might also be happening in other repositories.
@samjwu Can you take a look at the clang-format CI?

Run git fetch origin develop
From https://github.com/ROCm/rocRAND
 * branch            develop    -> FETCH_HEAD
 * [new branch]      develop    -> origin/develop
fatal: couldn't find remote ref develop_stream_20240701
Error: Process completed with exit code [12](https://github.com/ROCm/rocRAND/actions/runs/9746751218/job/26897847285?pr=512#step:6:13)8.

**Comment by Naraenda** [comment]
Rebased to get the clang-format ci fixes.

**Comment by Naraenda** [comment]
Rebased and added threefry fixes by @ex-rzr

**Review by umfranzw — APPROVED** [review]
I've finished working my way through this one, and I think it looks good. All our CI checks are also passing now.
@stanleytsang-amd if you have no objections, I'm ok with this one going in.

File: CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +33,7 @@ cmake_dependent_option(BUILD_BENCHMARK_TUNING
 option(BUILD_ADDRESS_SANITIZER "Build with address sanitizer enabled" OFF)
 option(CODE_COVERAGE "Build with code coverage flags (clang only)" OFF)
 option(DEPENDENCIES_FORCE_DOWNLOAD "Don't search the system for dependencies, always download them" OFF)
+cmake_dependent_option(RUN_SLOW_TESTS "Run extra tests with CTest. These cover niche functionality and take long time" OFF "BUILD_TEST" OFF)
```
Comment (stanleytsang-amd) [inline CMakeLists.txt:36]: Please update the packaging version in CMakeLists.txt to match the changelog.

**Review by stanleytsang-amd — CHANGES_REQUESTED** [review]
LGTM once the packaging version is updated.

File: CMakeLists.txt (hunk 1 above)
Comment (Naraenda) [reply CMakeLists.txt:36]: Done in ec409a3

Comment (stanleytsang-amd) [reply CMakeLists.txt:36]: thanks

**Review by stanleytsang-amd — APPROVED** [review]

**Pull request merged**